File: sdk/python/protocol_7h3/keys.py
"""Key infrastructure for 7h3 Protocol — discovery, rotation, revocation."""
from __future__ import annotations
import json
import time
import threading
from typing import Dict, List, Optional, Any
from dataclasses import dataclass, field

WELL_KNOWN_PATH = "/.well-known/7h3-keys"
REVOCATION_PATH = "/.well-known/7h3-revoked"

@dataclass
class KeyEntry:
    id: str
    algorithm: str  # 'Ed25519'
    public_key: str  # SPKI base64url
    created: int     # Unix ms
    expires: Optional[int] = None
    revoked: bool = False
    revoked_at: Optional[int] = None

@dataclass
class WellKnownKeysDocument:
    version: str  # '7h3/0.1'
    updated: int
    keys: List[KeyEntry]

    def to_json(self) -> str:
        keys_list = []
        for k in self.keys:
            entry = {
                "id": k.id, "algorithm": k.algorithm, "publicKey": k.public_key,
                "created": k.created,
            }
            if k.expires is not None: entry["expires"] = k.expires
            if k.revoked: entry["revoked"] = True
            if k.revoked_at is not None: entry["revokedAt"] = k.revoked_at
            keys_list.append(entry)
        return json.dumps({"version": self.version, "updated": self.updated, "keys": keys_list}, separators=(",", ":"))

    @classmethod
    def from_json(cls, data: str) -> "WellKnownKeysDocument":
        d = json.loads(data)
        keys = []
        for k in d.get("keys", []):
            keys.append(KeyEntry(
                id=k["id"], algorithm=k["algorithm"], public_key=k["publicKey"],
                created=k["created"], expires=k.get("expires"), revoked=k.get("revoked", False),
                revoked_at=k.get("revokedAt"),
            ))
        return cls(version=d["version"], updated=d["updated"], keys=keys)


def fetch_well_known_keys(base_url: str, *, timeout: float = 5.0) -> WellKnownKeysDocument:
    """Fetch the /.well-known/7h3-keys document from a base URL."""
    import urllib.request
    url = base_url.rstrip("/") + WELL_KNOWN_PATH
    with urllib.request.urlopen(url, timeout=timeout) as resp:
        data = resp.read().decode("utf-8")
    return WellKnownKeysDocument.from_json(data)


@dataclass
class ManagedKeyPair:
    id: str
    public_key: str   # SPKI base64url
    private_key: str  # PKCS8 base64url
    created: int
    expires_at: Optional[int] = None


class KeyRotationManager:
    """Manages Ed25519 key pairs with automatic rotation."""

    def __init__(self, max_age_ms: int = 86_400_000, overlap_ms: int = 3_600_000):
        self.max_age_ms = max_age_ms
        self.overlap_ms = overlap_ms
        self._keys: List[ManagedKeyPair] = []
        self._lock = threading.Lock()

    def add_key(self, pair: ManagedKeyPair) -> None:
        with self._lock:
            self._keys.append(pair)

    def get_current_key(self) -> Optional[ManagedKeyPair]:
        """Return the most recently created non-expired key."""
        now = int(time.time() * 1000)
        with self._lock:
            active = [k for k in self._keys if not k.expires_at or k.expires_at > now]
            if not active:
                return None
            return sorted(active, key=lambda k: k.created, reverse=True)[0]

    def rotate_if_needed(self) -> Optional[ManagedKeyPair]:
        """Generate a new key if the current one is too old.

        NOTE: generate_ed25519_keypair does not exist in protocol.py.
        Callers should supply keys externally via add_key() rather than
        relying on this method.
        """
        raise NotImplementedError(
            "rotate_if_needed requires generate_ed25519_keypair which is not "
            "implemented in protocol.py. Supply new keys externally via add_key()."
        )

    def get_well_known_document(self) -> WellKnownKeysDocument:
        now = int(time.time() * 1000)
        with self._lock:
            entries = []
            for k in self._keys:
                entry = KeyEntry(
                    id=k.id, algorithm="Ed25519", public_key=k.public_key,
                    created=k.created, expires=k.expires_at,
                    revoked=bool(k.expires_at and k.expires_at < now),
                )
                entries.append(entry)
        return WellKnownKeysDocument(version="7h3/0.1", updated=now, keys=entries)


class RevocationRegistry:
    """Tracks revoked key IDs."""

    def __init__(self):
        self._revoked: Dict[str, Dict[str, Any]] = {}
        self._lock = threading.Lock()

    def revoke(self, key_id: str, reason: Optional[str] = None) -> None:
        with self._lock:
            self._revoked[key_id] = {"revokedAt": int(time.time() * 1000), "reason": reason}

    def is_revoked(self, key_id: str) -> bool:
        with self._lock:
            return key_id in self._revoked

    def get_list(self) -> dict:
        now = int(time.time() * 1000)
        with self._lock:
            revoked = [
                {"id": kid, "revokedAt": v["revokedAt"], **({"reason": v["reason"]} if v["reason"] else {})}
                for kid, v in self._revoked.items()
            ]
        return {"version": "7h3/0.1", "updated": now, "revokedKeys": revoked}

    def import_list(self, revocation_list: dict) -> None:
        with self._lock:
            for entry in revocation_list.get("revokedKeys", []):
                kid = entry["id"]
                if kid not in self._revoked:
                    self._revoked[kid] = {"revokedAt": entry["revokedAt"], "reason": entry.get("reason")}

File: sdk/rust/src/keys.rs
//! Key infrastructure for 7h3 Protocol — discovery, rotation, revocation.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};
use zeroize::{Zeroize, ZeroizeOnDrop};

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

/// An entry in the /.well-known/7h3-keys document.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KeyEntry {
    pub id: String,
    pub algorithm: String,    // "Ed25519"
    #[serde(rename = "publicKey")]
    pub public_key: String,   // SPKI base64url
    pub created: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expires: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub revoked: Option<bool>,
    #[serde(rename = "revokedAt", skip_serializing_if = "Option::is_none")]
    pub revoked_at: Option<u64>,
}

/// The /.well-known/7h3-keys document.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WellKnownKeysDocument {
    pub version: String,  // "7h3/0.1"
    pub updated: u64,
    pub keys: Vec<KeyEntry>,
}

impl WellKnownKeysDocument {
    pub fn new(keys: Vec<KeyEntry>) -> Self {
        Self { version: "7h3/0.1".to_string(), updated: now_ms(), keys }
    }

    pub fn to_json(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(self)
    }

    pub fn from_json(s: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(s)
    }

    /// Find an active (non-revoked, non-expired) public key by ID.
    pub fn get_active_key(&self, key_id: &str) -> Option<&KeyEntry> {
        let now = now_ms();
        self.keys.iter().find(|k| {
            k.id == key_id
                && !k.revoked.unwrap_or(false)
                && k.expires.map_or(true, |e| e > now)
        })
    }
}

/// Managed key pair for rotation.
///
/// The private key is wiped from memory on drop, and the `Debug` impl
/// redacts it so key material can never reach logs via `{:?}`.
#[derive(Clone, Zeroize, ZeroizeOnDrop)]
pub struct ManagedKeyPair {
    #[zeroize(skip)]
    pub id: String,
    #[zeroize(skip)]
    pub public_key: String,   // SPKI base64url
    pub private_key: String,  // PKCS8 base64url
    #[zeroize(skip)]
    pub created: u64,
    #[zeroize(skip)]
    pub expires_at: Option<u64>,
}

impl std::fmt::Debug for ManagedKeyPair {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ManagedKeyPair")
            .field("id", &self.id)
            .field("public_key", &self.public_key)
            .field("private_key", &"<redacted>")
            .field("created", &self.created)
            .field("expires_at", &self.expires_at)
            .finish()
    }
}

/// Manages key pairs with rotation policy.
pub struct KeyRotationManager {
    pub max_age_ms: u64,
    pub overlap_ms: u64,
    keys: Arc<Mutex<Vec<ManagedKeyPair>>>,
}

impl KeyRotationManager {
    pub fn new(max_age_ms: u64, overlap_ms: u64) -> Self {
        Self {
            max_age_ms,
            overlap_ms,
            keys: Arc::new(Mutex::new(Vec::new())),
        }
    }

    pub fn add_key(&self, pair: ManagedKeyPair) {
        self.keys.lock().unwrap().push(pair);
    }

    /// Returns the most recently created non-expired key.
    pub fn get_current_key(&self) -> Option<ManagedKeyPair> {
        let now = now_ms();
        let keys = self.keys.lock().unwrap();
        keys.iter()
            .filter(|k| k.expires_at.map_or(true, |e| e > now))
            .max_by_key(|k| k.created)
            .cloned()
    }

    /// Build a WellKnownKeysDocument from the managed keys.
    pub fn get_well_known_document(&self) -> WellKnownKeysDocument {
        let now = now_ms();
        let keys = self.keys.lock().unwrap();
        let entries: Vec<KeyEntry> = keys.iter().map(|k| KeyEntry {
            id: k.id.clone(),
            algorithm: "Ed25519".to_string(),
            public_key: k.public_key.clone(),
            created: k.created,
            expires: k.expires_at,
            revoked: if k.expires_at.map_or(false, |e| e < now) { Some(true) } else { None },
            revoked_at: None,
        }).collect();
        WellKnownKeysDocument::new(entries)
    }
}

/// Tracks revoked key IDs in memory.
#[derive(Default)]
pub struct RevocationRegistry {
    revoked: Arc<Mutex<HashMap<String, RevocationEntry>>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RevocationEntry {
    #[serde(rename = "revokedAt")]
    pub revoked_at: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

impl RevocationRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn revoke(&self, key_id: &str, reason: Option<&str>) {
        self.revoked.lock().unwrap().insert(
            key_id.to_string(),
            RevocationEntry { revoked_at: now_ms(), reason: reason.map(String::from) },
        );
    }

    pub fn is_revoked(&self, key_id: &str) -> bool {
        self.revoked.lock().unwrap().contains_key(key_id)
    }

    pub fn get_list(&self) -> serde_json::Value {
        let revoked = self.revoked.lock().unwrap();
        let entries: Vec<serde_json::Value> = revoked.iter().map(|(id, entry)| {
            let mut obj = serde_json::json!({ "id": id, "revokedAt": entry.revoked_at });
            if let Some(ref r) = entry.reason {
                obj["reason"] = serde_json::json!(r);
            }
            obj
        }).collect();
        serde_json::json!({ "version": "7h3/0.1", "updated": now_ms(), "revokedKeys": entries })
    }
}
